scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,268 @@
             sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            is_group = True if subject_rec["other_config"]["subject-is-group"] is "true" else False
+            logger.debug("Permit %s with sid %s is_group as %s", name, sid, is_group)
             if admin_role in subject_rec['roles']:
-                config.add_subject(sid)
-        config.install()
+                self._add_subject(name, is_group)
+
+    def _add_subject(self, name, is_group):
+        condition = " ingroup {}".format(name) if is_group else " = {}".format(name)
+        self._lines.append("account sufficient pam_succeed_if.so user {} \n".format(condition))
+
+    def _install(self):
+        if self._ad_enabled:
+            super(DynamicPam, self)._install()
+        else:
+            if os.path.exists(self._file_path):
+                os.remove(self._file_path)
+
+
+class KeyValueConfig(ADConfig):
+    def __init__(self, path, session, args, ad_enabled=True, sep=": "):
+        super(KeyValueConfig, self).__init__(path, session, args, ad_enabled)
+        self._sep = sep
+        self._values = {}
+
+    def _update_value(self, key, value):
+        self._values[key] = value
+
+    def _apply_value(self, key, value):
+        try:
+            target_line = "{}{}{}\n".format(key, self._sep, value)
+            for idx, line in enumerate(self._lines):
+                if key not in line:
+                    continue
+                kv =  line.split(self._sep)
+                k , v = kv[0].strip(), kv[1].strip()
+                if k != key:
+                    continue
+                self._lines[idx] = target_line
+                return
+            # Does not existed, append to the end
+            self._lines.append(target_line)
+        except Exception:
+            logger.exception("Failed to update %s to %s", key, value)
+
+    def _apply_to_cache(self):
+        for k, v in self._values.items():
+            self._apply_value(k, v)
+
+
+class NssConfig(KeyValueConfig):
+    def __init__(self, session, args, ad_enabled=True):
+        super(NssConfig, self).__init__("/etc/nsswitch.conf", session, args, ad_enabled)
+        modules = "files sss"
+        if ad_enabled:
+            if self._backend == ADBackend.bd_pbis:
+                modules = "files sss lsass"
+            else:
+                modules = "files hcp winbind"
+        self._update_value("passwd", modules)
+        self._update_value("group", modules)
+        self._update_value("shadow", modules)
+
+
+class SshdConfig(KeyValueConfig):
+    def __init__(self, session, args, ad_enabled=True):
+        super(SshdConfig, self).__init__("/etc/ssh/sshd_config", session, args, ad_enabled, sep=" ")
+        value = "yes" if ad_enabled else "no"
+        self._values["ChallengeResponseAuthentication"] = value
+        self._values["GSSAPIAuthentication"] = value
+        self._values["GSSAPICleanupCredentials"] = value
```

Comment:
You abstracted away the `self._values` dictionary in the `KeyValueConfig` class with `def _update_value`, so you should be consistent here